### PR TITLE
Encapsulate project with Docker for development and production

### LINKED ISSUE #62: Add Verto as dependency when released to PyPI

### LINKED ISSUE #174: Encapsulate project with Docker for development
## Description

The project should be easily installed, developed, and deployed by encapsulating the project within a Docker image.

### To do

- **Development**
  - [x] Create `docker-compose.yml` for developers/authors.
    - [x] Create and connect nginx container with Node for static files
      - [x] Check creation of Scratch images
    - [x] Create and connect Postgres database container
    - [x] Create and connect to gunicorn + Django container
    - [x] Add Weasyprint to Django container or create and connect separate Weasyprint container
      - Weasyprint has now been added within the Django container.
      - ~~*Note:* This was implemented with the following caveats:~~
        - ~~The method of creating PDFs now uses wkhtmltopdf and pdfkit instead of Weasyprint as it's easier to setup dependencies. In the long term, it would be good to switch back to Weasyprint due to the features it offers.~~
        - ~~The PDF generation scripts searches for static files, located in the build folder. These files are only available because the build folder is mounted onto both the nginx and django containers. In production, the static files would not be on the same server as the Django instance. Will need to think about possible solutions for production.~~
        - ~~An issue with wkhtmltopdf was referencing a local image for the footer. The image has been removed for the development version.~~
        - ~~Due to issues with wkhtmltopdf, the header text has been removed. Also the footer text does not have a custom colour or font applied.~~
  - [x] Create helper script aimed for authors but also helpful for developers
  - [ ] Fix Django Debug Toolbar to display correctly
  - [x] Display static files for admin website Fixed in a55d63fd1269b715e30deb5d0350f5f9552198ea.
  - [x] Fix bug: Scratch images are missing hash in filename. Fixed in f7715752cde87060ad7abc5dada549a6974d5921.
- **Production**
  - Moved to #184
- **Cleanup**
  - [x] Sort out `.env` file - Left as is as required for production.
  - [x] Write documentation

### To research

- [ ] Should `Dockerfile` be moved to be within `csunplugged` folder?
- [ ] Use of `.dockerignore`
- [ ] Figure out to access Django errors (can create one when trying to create PDF)
- [x] Figure out `ENV DEBIAN_FRONTEND noninteractive` is used for? (now against best practices?)
  - Already enabled in base Google image
- [x] If `docker build .` takes several minutes to build, does this mean GAE flex instances will take time to boot up as well?
  - Build would only happen once, it's the container creation time we need to be aware of.

### Useful links

- https://cloud.google.com/appengine/docs/flexible/custom-runtimes/

*Note: While this issue relates to every milestone, it is connected to the 'Ready for Authors' milestone as it changes the way authors use the system. We want to establish the author/developer environment before we train them in the system.*

### PR DESCRIPTION
**Pull request by JackMorganNZ**
The project should be easily installed, developed, and deployed by encapsulating the project within a Docker image.

The following changes were made:

- Create helper script aimed for authors but also helpful for developers for using new system.
- Use Docker and Docker compose for local development that emulates parts of the production environment.
- Use Docker for Google App Engine deployment for production.
- Updated documentation for installation, local development, and manual deployment.
- Change CI to use Docker system.

Full details are available here: #174 #184.

This pull request also fixes #62.

Issues that need to be raised when PR is accepted:

- Fix Django Debug Toolbar to display correctly.
- Check Docker image for size efficiency.
- Check Docker image for speed efficiency.
- Lock Docker image to specific version.
- Update documentation flowcharts with correct command (after #188 is merged).
- Simplify Gulp script.
- Automate deployment (using production build of static files).

### CODE REVIEW

**Comment by hayleyavw** [comment]
```
ERROR: In file './docker-compose.yml' service 'version' doesn't have any configuration options. All top level keys in your docker-compose.yml must map to a dictionary of configuration options.
Waiting until all systems are ready... (takes roughly 10 to 20 seconds)
```
Running `./csu start` on my ubuntu vm gave this ^ output. Been going for about 45mins at time of commenting, will wait and see if it works before approving.... In the mean time, will test it out on my laptop.

**Comment by hayleyavw** [comment]
Update: laptop had been working on `./csu start` for two hours this morning when I left it, not sure if it will ever finish....
Currently got `./csu start` running on ubuntu VM, but had to use these steps:
```
sudo -i
curl -L https://github.com/docker/compose/releases/download/1.12.0/docker-compose-`uname -s`-`uname -m` > /usr/local/bin/docker-compose
chmod +x /usr/local/bin/docker-compose
exit
sudo ./csu start
```
It didn't like the version of docker-compose, and could only install the newer version (1.12.0) through the superuser.
I used this page as reference: https://docs.docker.com/compose/install/

File: docs/source/getting_started/installation.rst [hunk 1]
```diff
@@ -5,220 +5,175 @@ This page will set your machine up for working on the CS Unplugged project.
...
 .. code-block:: bash
 
-    $ createuser --interactive
-    $ Enter name of role to add: <your name>
-    $ Shall the new role be a superuser? (y/n): y
-
-For example:
+    $ docker version
+    $ docker-compose version
+    $ docker run hello-world
```
Comment (JackMorganNZ) [inline docs/source/getting_started/installation.rst]: Will add following note:

> Depending on your operating system, if the above commands don't work you
> may need to set Docker to be able to run without ``sudo``, in order to use
> the ``csu`` helper script.

File: docs/source/getting_started/installation.rst [hunk 2]
```diff
@@ -5,220 +5,175 @@ This page will set your machine up for working on the CS Unplugged project.
...
 
 .. code-block:: bash
 
-    $ createuser --interactive
-    $ Enter name of role to add: <your name>
-    $ Shall the new role be a superuser? (y/n): y
-
-For example:
+    $ docker version
+    $ docker-compose version
```
Comment (JackMorganNZ) [inline docs/source/getting_started/installation.rst]: Not expected, did you install Docker through the Docker store?

**Comment by JackMorganNZ** [comment]
I'm unsure if your comments are due to the custom installations of Docker, as it worked first time on Tracy's machine. I need to wipe your laptop of Docker and try again.

**Comment by JackMorganNZ** [comment]
Ping @hayleyavw for recheck.

**Comment by JackMorganNZ** [comment]
Ping @hayleyavw for recheck

**Comment by ravenmaster001** [comment]
Same error as @hayleyavw had with the incorrect docker-compose version. I'm using Fedora as my OS and it seems docker-compose is only 1.9.0 and docker is 1.27 from the docker-store repository. Otherwise I'm waiting for it to pull images.

Error Message:
```
Creating systems...
ERROR: Version in "./docker-compose.yml" is unsupported. You might be seeing this error because you're using the wrong Compose file version. Either specify a version of "2" (or "2.0") and place your service definitions under the `services` key, or omit the `version` key and place your service definitions at the root of the file to use version 1.
For more on the Compose file format versions, see https://docs.docker.com/compose/compose-file/
Waiting until all systems are ready... (takes roughly 10 to 20 seconds)
```

**Comment by JackMorganNZ** [comment]
@ravenmaster001 The waiting step will never finish (I think) and I'm not clever enough to know how to add a check. Can you update `docker-compose` and try it again.

**Pull request merged**